### ISSUE TITLE
how to train a "mask keypoint r-cnn"

### ISSUE THREAD

**Issue by feiyangsuo** [pytorch/pytorch#48927]
**Question**
To have a detection model predict bbox, mask and keypoints simultaneously, I wrote a script of "mask keypoint r-cnn", based on pytorch's indigenous implementation of  mask r-cnn and keypoint r-cnn.

To test the baseline I use a dataset with 10 images of pedestrians, labeled with keypoints and masks. In training I sum the losses of each module together and optimize it. But the result is unsatisfactory after even 200 epochs. Neither the predicted bboxes nor keypoints and masks looks fine. Yet the loss seems already converged and no more decreasing.

In my expectation, with so few samples, it should be easy for the model to overfit the dataset.

I tried ignoring one among mask loss and keypoint loss, then the model is well trained as expected, becoming a good keypoint r-cnn, or mask r-cnn. I think this proves my implementation didn't go wrong.

The question, is there advise or experience for training keypoint and mask together? Thanks in advance :)

**Appendix**
My implementation of mask keypoint r-cnn:
```
import torch
from torchvision.models.utils import load_state_dict_from_url
from torchvision.ops import MultiScaleRoIAlign
from torchvision.models.detection.faster_rcnn import FasterRCNN
from torchvision.models.detection.backbone_utils import resnet_fpn_backbone
from torchvision.models.detection.mask_rcnn import MaskRCNNHeads, MaskRCNNPredictor
from torchvision.models.detection.keypoint_rcnn import KeypointRCNNHeads, KeypointRCNNPredictor
import time


class MaskKeypointRCNN(FasterRCNN):
    def __init__(self, backbone, num_classes=None,
                 # transform parameters
                 min_size=800, max_size=1333,
                 image_mean=None, image_std=None,
                 # RPN parameters
                 rpn_anchor_generator=None, rpn_head=None,
                 rpn_pre_nms_top_n_train=2000, rpn_pre_nms_top_n_test=1000,
                 rpn_post_nms_top_n_train=2000, rpn_post_nms_top_n_test=1000,
                 rpn_nms_thresh=0.7,
                 rpn_fg_iou_thresh=0.7, rpn_bg_iou_thresh=0.3,
                 rpn_batch_size_per_image=256, rpn_positive_fraction=0.5,
                 # Box parameters
                 box_roi_pool=None, box_head=None, box_predictor=None,
                 box_score_thresh=0.05, box_nms_thresh=0.5, box_detections_per_img=100,
                 box_fg_iou_thresh=0.5, box_bg_iou_thresh=0.5,
                 box_batch_size_per_image=512, box_positive_fraction=0.25,
                 bbox_reg_weights=None,
                 # Mask parameters
                 mask_roi_pool=None, mask_head=None, mask_predictor=None,
                 # keypoint parameters
                 keypoint_roi_pool = None, keypoint_head = None, keypoint_predictor = None,
                 num_keypoints = 17):

        out_channels = backbone.out_channels

        # mask predictor initialization
        assert isinstance(mask_roi_pool, (MultiScaleRoIAlign, type(None)))
        if num_classes is not None:
            if mask_predictor is not None:
                raise ValueError("num_classes should be None when mask_predictor is specified")
        if mask_roi_pool is None:
            mask_roi_pool = MultiScaleRoIAlign(
                featmap_names=['0', '1', '2', '3'],
                output_size=14,
                sampling_ratio=2)
        if mask_head is None:
            mask_layers = (256, 256, 256, 256)
            mask_dilation = 1
            mask_head = MaskRCNNHeads(out_channels, mask_layers, mask_dilation)
        if mask_predictor is None:
            mask_predictor_in_channels = 256  # == mask_layers[-1]
            mask_dim_reduced = 256
            mask_predictor = MaskRCNNPredictor(mask_predictor_in_channels,
                                               mask_dim_reduced, num_classes)

        # keypoint predictor initialization
        assert isinstance(keypoint_roi_pool, (MultiScaleRoIAlign, type(None)))
        if min_size is None:
            min_size = (640, 672, 704, 736, 768, 800)
        if num_classes is not None:
            if keypoint_predictor is not None:
                raise ValueError("num_classes should be None when keypoint_predictor is specified")
        if keypoint_roi_pool is None:
            keypoint_roi_pool = MultiScaleRoIAlign(
                featmap_names=['0', '1', '2', '3'],
                output_size=14,
                sampling_ratio=2)
        if keypoint_head is None:
            keypoint_layers = tuple(512 for _ in range(8))
            keypoint_head = KeypointRCNNHeads(out_channels, keypoint_layers)
        if keypoint_predictor is None:
            keypoint_dim_reduced = 512  # == keypoint_layers[-1]
            keypoint_predictor = KeypointRCNNPredictor(keypoint_dim_reduced, num_keypoints)

        super(MaskKeypointRCNN, self).__init__(
            backbone, num_classes,
            # transform parameters
            min_size, max_size,
            image_mean, image_std,
            # RPN-specific parameters
            rpn_anchor_generator, rpn_head,
            rpn_pre_nms_top_n_train, rpn_pre_nms_top_n_test,
            rpn_post_nms_top_n_train, rpn_post_nms_top_n_test,
            rpn_nms_thresh,
            rpn_fg_iou_thresh, rpn_bg_iou_thresh,
            rpn_batch_size_per_image, rpn_positive_fraction,
            # Box parameters
            box_roi_pool, box_head, box_predictor,
            box_score_thresh, box_nms_thresh, box_detections_per_img,
            box_fg_iou_thresh, box_bg_iou_thresh,
            box_batch_size_per_image, box_positive_fraction,
            bbox_reg_weights)

        self.roi_heads.mask_roi_pool = mask_roi_pool
        self.roi_heads.mask_head = mask_head
        self.roi_heads.mask_predictor = mask_predictor
        self.roi_heads.keypoint_roi_pool = keypoint_roi_pool
        self.roi_heads.keypoint_head = keypoint_head
        self.roi_heads.keypoint_predictor = keypoint_predictor
```

My training script:
```
def train_and_eval(model,
                   dataloader_train,
                   epochs, batches_show=10,
                   lr=0.01,
                   keypoint_enabled=True, mask_enabled=True,
                   save_dir=None, save_interval=10,
                   load_from=None):
    if torch.cuda.is_available():
        device = torch.device("cuda")
    else:
        device = torch.device("cpu")
        print("CUDA is unavailable, using CPU instead.")

    model = model.to(device)
    if load_from is not None:
        model.load_state_dict(torch.load(load_from))

    params = [p for p in model.parameters() if p.requires_grad]
    optimizer = torch.optim.SGD(params, lr=lr, momentum=0.9, weight_decay=1e-6)
    lr_scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, factor=0.1, patience=5, verbose=True)

    model.train()
    for epoch in range(1, epochs+1):
        print("----------------------  TRAINING  ---------------------- ")
        running_losses = {}
        for i, data in enumerate(dataloader_train, start=1):
            images, targets = data
            images = [image.to(device) for image in images]
            targets = [{k: v.to(device) for k, v in t.items()} for t in targets]

            loss_dict = model(images, targets)
            if not keypoint_enabled:
                loss_dict.pop("loss_keypoint")
            if not mask_enabled:
                loss_dict.pop("loss_mask")
            loss = sum(loss for loss in loss_dict.values())

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            for k, v in loss_dict.items():
                if k in running_losses:
                    running_losses[k] += v.item()
                else:
                    running_losses[k] = v.item()

            if i % batches_show == 0:
                running_losses = {k: v / batches_show for k, v in running_losses.items()}
                running_loss = sum(v for v in running_losses.values())
                summary = f'[epoch: {epoch}, batch: {i}] [loss: {running_loss:.3f}] ' + " ".join([f"[{k}: {v:.3f}]" for k, v in running_losses.items()])
                print(summary)
                running_losses = {}
        lr_scheduler.step(loss)

        if save_dir is not None:
            os.makedirs(save_dir, exist_ok=True)
            if epoch % save_interval == 0:
                print("----------------------   SAVING   ---------------------- ")
                # torch.save(model, os.path.join(save_dir, "epoch_{}.pth".format(epoch)))
                torch.save(model.state_dict(), os.path.join(save_dir, "epoch_{}.state_dict.pth".format(epoch)))
```

**Comment by ailzhang** [comment]
Hi! As instructed in the issue template, please reopen the questions in pytorch forum, thanks!

**Comment by feiyangsuo** [comment]
The problem seems solved after I tuned the weight of keypoint loss to 0.1.

(Is this kind of problems just things about parameter adjusting?)
